src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -369,18 +351,27 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
 
         protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         {
-            if (_returnTypesOpt != null)
+            if (!_returnTypesOpt.IsDefault)
             {
-                _returnTypesOpt.Clear();
+                _returnTypesOpt = _returnTypesOpt.Clear();
             }
             this.Diagnostics.Clear();
             ParameterSymbol methodThisParameter = MethodThisParameter;
             this.regionPlace = RegionPlace.Before;
-            EnterParameters(); // assign parameters
+            var parametersBuilder = PooledDictionary<Symbol, TypeWithAnnotations>.GetInstance();
+            foreach (var (key, value) in _variableTypes)
+            {
+                parametersBuilder.Add(key, value);
+            }
+            EnterParameters(parametersBuilder); // assign parameters
             if (!(methodThisParameter is null))
             {
-                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations);
+                EnterParameter(methodThisParameter, methodThisParameter.TypeWithAnnotations, parametersBuilder);
             }
+            _variableTypes = parametersBuilder.ToImmutableDictionaryAndFree();
+
+            // We need to do a checkpoint even of the first node, because we want to have the state of the initial parameters.
+            _snapshotter.SnapshotWalker(methodMainNode);
```

Comment:
>_snapshotter.SnapshotWalker(methodMainNode); [](start = 12, length = 44)

When we are taking a snapshot for a bound node, it it a snapshot of a state before we visit the node or after the node has been visited? In any case, I do not feel comfortable about this special case for the root bound node.

Reply:
It's the state before the node. With us moving to position based, are you still uncomfortable with this? We need to take a snapshot of the initial state of the parameters at the start of the method body .

---
In reply to: [288784061](https://github.com/dotnet/roslyn/pull/35850#discussion_r288784061) [](ancestors = 288784061)

Reply:
> With us moving to position based, are you still uncomfortable with this?

I think position-wise that would make sense.

---
In reply to: [288815668](https://github.com/dotnet/roslyn/pull/35850#discussion_r288815668) [](ancestors = 288815668,288784061)

Reply:
Right, and the first position of the method is the main body node, at which point all of the state we just set above (namely the initial states of the parameters) will need to be populated in a snapshot.

---
In reply to: [289073267](https://github.com/dotnet/roslyn/pull/35850#discussion_r289073267) [](ancestors = 289073267,288815668,288784061)